jxl/src/headers/permutation.rs

```diff
@@ -0,0 +1,336 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+use crate::bit_reader::BitReader;
+use crate::entropy_coding::decode::Reader;
+use crate::error::{Error, Result};
+use crate::util::value_of_lowest_1_bit;
+use crate::util::CeilLog2;
+
+pub struct Permutation(Vec<u32>);
+
+impl std::ops::Deref for Permutation {
+    type Target = [u32];
+
+    fn deref(&self) -> &[u32] {
+        &self.0
+    }
+}
+
+impl Permutation {
+    /// Decode a permutation from entropy-coded stream.
+    pub fn decode(
+        size: u32,
+        skip: u32,
+        br: &mut BitReader,
+        entropy_reader: &mut Reader,
+    ) -> Result<Self> {
+        let end = entropy_reader.read(br, get_context(size))?;
+        Self::decode_inner(size, skip, end, |ctx| entropy_reader.read(br, ctx))
+    }
+
+    fn decode_inner(
+        size: u32,
+        skip: u32,
+        end: u32,
+        mut read: impl FnMut(usize) -> Result<u32>,
+    ) -> Result<Self> {
+        if end > size - skip {
+            return Err(Error::InvalidPermutationSize { size, skip, end });
+        }
+
+        let mut lehmer = Vec::new();
+        lehmer.try_reserve(end as usize)?;
+
+        let mut prev_val = 0u32;
+        for idx in skip..(skip + end) {
+            let val = read(get_context(prev_val))?;
+            if val >= size - idx {
+                return Err(Error::InvalidPermutationLehmerCode {
+                    size,
+                    idx,
+                    lehmer: val,
+                });
+            }
+            lehmer.push(val);
+            prev_val = val;
+        }
+
+        // Initialize the full permutation vector with skipped elements intact
+        let mut permutation: Vec<u32> = Vec::new();
+        permutation.try_reserve((size - skip) as usize)?;
+        permutation.extend(0..size);
+
+        // Decode the Lehmer code into the slice starting at `skip`
+        let permuted_slice = decode_lehmer_code(&lehmer, &permutation[skip as usize..])?;
+
+        // Replace the target slice in `permutation`
+        permutation.splice(skip as usize.., permuted_slice);
```

Comment:
This could be `permutation[skip..].copy_from_slice(...)` (equivalent of `memcpy`), though they might compile to the same machine code.